universe/logic/api/auth.go

```diff
@@ -83,6 +85,38 @@ func VerifyPolkadotSignature(wallet, challenge, signature string) (bool, error)
 	return pub.Verify(sig, transcript)
 }
 
+func signHash(data []byte) []byte {
+	msg := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(data), data)
+	return crypto.Keccak256([]byte(msg))
+}
+
+func VerifyEthereumSignature(address, challenge, signature string) (bool, error) {
+	if !strings.HasPrefix(signature, "0x") {
+		signature = "0x" + signature
+	}
+	sigBytes := hexutil.MustDecode(signature)
+	if len(sigBytes) <= 64 || (sigBytes[64] != 27 && sigBytes[64] != 28) {
+		return false, errors.New("unsupported signature format")
+	}
+
+	sigBytes[64] -= 27
+
+	msgBytes := signHash([]byte(challenge))
+
+	pubKey, err := crypto.SigToPub(msgBytes, sigBytes)
+	if err != nil {
+		return false, errors.Wrap(err, "failed to recover public key")
+	}
+
+	recoveredAddr := crypto.PubkeyToAddress(*pubKey).Hex()
```

Comment:
Maybe check pubKey for nil before dereferencing pointer

Reply:
added this, 10x
https://github.com/momentum-xyz/ubercontroller/pull/102